File: apps/a-demo-be-rust-axum/src/handlers/test_api.rs
use axum::{extract::State, http::StatusCode, response::IntoResponse, Json};
use serde::Deserialize;
use serde_json::json;
use std::sync::Arc;

use crate::domain::errors::AppError;
use crate::state::AppState;

/// Deletes all data in dependency order:
/// attachments → expenses → refresh_tokens → revoked_tokens → users.
pub async fn reset_db(State(state): State<Arc<AppState>>) -> Result<impl IntoResponse, AppError> {
    sqlx::query("DELETE FROM attachments")
        .execute(&state.pool)
        .await?;
    sqlx::query("DELETE FROM expenses")
        .execute(&state.pool)
        .await?;
    sqlx::query("DELETE FROM refresh_tokens")
        .execute(&state.pool)
        .await?;
    sqlx::query("DELETE FROM revoked_tokens")
        .execute(&state.pool)
        .await?;
    sqlx::query("DELETE FROM users")
        .execute(&state.pool)
        .await?;

    Ok((
        StatusCode::OK,
        Json(json!({"message": "Database reset successful"})),
    ))
}

#[derive(Deserialize)]
pub struct PromoteAdminRequest {
    pub username: String,
}

/// Sets the role of the given user to "ADMIN".
/// Returns 404 if the username does not exist.
pub async fn promote_admin(
    State(state): State<Arc<AppState>>,
    Json(body): Json<PromoteAdminRequest>,
) -> Result<impl IntoResponse, AppError> {
    let row = sqlx::query("SELECT id FROM users WHERE username = $1")
        .bind(&body.username)
        .fetch_optional(&state.pool)
        .await?;

    let _row = row.ok_or_else(|| AppError::NotFound {
        entity: "user".to_string(),
    })?;

    sqlx::query("UPDATE users SET role = 'ADMIN' WHERE username = $1")
        .bind(&body.username)
        .execute(&state.pool)
        .await?;

    Ok((
        StatusCode::OK,
        Json(json!({"message": format!("User {} promoted to ADMIN", body.username)})),
    ))
}

File: apps/demo-be-python-fastapi/src/demo_be_python_fastapi/routers/test_api.py
"""Test API router: database reset and admin promotion endpoints.

These endpoints are only available when ENABLE_TEST_API=true.
They exist to support FE E2E test setup and teardown.
"""

import os

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from demo_be_python_fastapi.dependencies import get_db
from demo_be_python_fastapi.domain.errors import NotFoundError
from demo_be_python_fastapi.infrastructure.models import (
    AttachmentModel,
    ExpenseModel,
    RevokedTokenModel,
    UserModel,
)

router = APIRouter()


class PromoteAdminRequest(BaseModel):
    """Request body for promote-admin endpoint."""

    username: str


def _is_test_api_enabled() -> bool:
    return os.getenv("ENABLE_TEST_API") == "true"


@router.post("/reset-db")
def reset_db(db: Session = Depends(get_db)) -> dict[str, str]:
    """Delete all data in dependency order: attachments → expenses → revoked_tokens → users."""
    db.query(AttachmentModel).delete()
    db.query(ExpenseModel).delete()
    db.query(RevokedTokenModel).delete()
    db.query(UserModel).delete()
    db.commit()
    return {"message": "Database reset successful"}


@router.post("/promote-admin")
def promote_admin(
    body: PromoteAdminRequest,
    db: Session = Depends(get_db),
) -> dict[str, str]:
    """Set the given user's role to ADMIN."""
    from sqlalchemy import select

    user = db.execute(
        select(UserModel).where(UserModel.username == body.username)
    ).scalar_one_or_none()
    if user is None:
        raise NotFoundError(f"User {body.username} not found")
    user.role = "ADMIN"
    db.commit()
    return {"message": f"User {body.username} promoted to ADMIN"}
